src/lib.rs

```diff
@@ -40,19 +42,314 @@ mod repl_helper;
+        ///
+        #[no_mangle]
+        pub extern "C" fn scryer_machine_new() -> *mut Machine {
+            let machine = Box::into_raw(Box::new(Machine::new_lib()));
+            machine
+        }
+
+
+        /// Frees the memory occupied by a [`Machine`] object.
+        ///
+        /// # Safety
+        ///
+        /// * It is the caller's responsibility to ensure that the input is a valid [`Machine`] pointer
+        /// created by [`scryer_machine_new`] that has not yet been freed by this function.
+        /// * Any [`QueryState`] instances associated with this machine should be deallocated prior
+        /// to calling this function on the input, otherwise those [`QueryState`] instances may
+        /// not be deallocated until the program terminates.
+        ///
+        ///
+        /// # Parameters
+        ///
+        /// - `ptr`: A mutable raw pointer to a `Machine` object.
+        ///
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_machine_free(ptr: *mut Machine) {
+            unsafe { drop(Box::from_raw(ptr)); }
+        }
+
+        /// Returns a new query generator for the given virtual machine.
+        ///
+        /// # Safety
+        ///
+        /// Caller must satisfy the following preconditions for this function:
+        ///
+        /// * Valid [`Machine`] pointer created with [`scryer_machine_new`] that has not yet been freed.
+        /// * an existing allocated [`QueryState`] for *this* [`Machine`] must not already exist.
+        /// * the input must represent valid Scryer Prolog code which fulfils the safety requirements of [`CStr::From_ptr`].
+        /// * There must be no other [`QueryState`] for this [`Machine`] started by 
+        /// [`scryer_start_new_query_generator`] that has not yet been freed with 
+        /// [`scryer_cleanup_query_generator`] or the [`Machine`] state will enter an undefined 
+        /// configuration with unpredictable results.  
+        ///
+        /// Other concerns:
+        /// * after invoking this function, calling any other function besides [`scryer_run_query_generator`]
+        /// before invoking [`scryer_cleanup_query_generator`] on the [`Machine`] pointer and [`QueryState`] pointer
+        /// will leave the [`Machine`] in an undefined state
+        ///
+        /// # Arguments
+        ///
+        /// * `machine` - A mutable reference to a [`Machine`] object.
+        /// * `input` - A pointer to a null terminated UTF-8 string.
+        ///
+        /// # Returns
+        ///
+        /// If the input string is successfully converted to a Rust string, the function returns a mutable pointer to a [`QueryState`] object.
+        /// If the input string is not a valid UTF-8 string or an error occurs during the query generation, the function returns a null pointer.
+        ///
+        #[no_mangle]
+        pub unsafe extern "C" fn scryer_start_new_query_generator(machine: &mut Machine, input: *const c_char) -> *mut QueryState {
```

Comment:
I think it's a good time to bikeshed on the names of the functions. I will give my suggestions in the spirit of what these interfaces should look in the Rust side after #2472.

```suggestion
        pub unsafe extern "C" fn scryer_run_query_iter(machine: &mut Machine, input: *const c_char) -> *mut QueryState {
```